source/common/router/config_impl.cc

```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
       const auto result = match.result_();
       if (result->typeUrl() == RouteMatchAction::staticTypeUrl()) {
         const RouteMatchAction& route_action = result->getTyped<RouteMatchAction>();
+        if (!headers.Path() && !route_action.route()->supportsPathlessHeaders()) {
+          return nullptr;
+        }
 
-        if (route_action.route()->matches(headers, stream_info, random_value)) {
-          return route_action.route();
+        RouteConstSharedPtr route_entry =
+            route_action.route()->matches(headers, stream_info, random_value);
+        if (nullptr == route_entry) {
+          ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
+          return nullptr;
         }
 
-        ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
-        return nullptr;
+        if (cb) {
+          if (cb(route_entry, RouteEvalStatus::NoMoreRoutes) == RouteMatchStatus::Accept) {
+            return route_entry;
+          }
+          return nullptr;
+        }
+
+        return route_entry;
```

Comment:
The code block here (from 1882 to 1900) can also  be replaced with helper function I suggested below.  Because it is just the case where route vector has 1 element (i.e., {route_action.route()})

Basically, we replace these three big code blocks to three  one-line func calls to the helper function.